distributed/recreate_tasks.py

```diff
@@ -0,0 +1,172 @@
+from .worker import _deserialize
+
+logger = logging.getLogger(__name__)
+
+
+class ReplayTaskScheduler:
+    """A plugin for the scheduler to recreate tasks locally
+
+    This adds the following routes to the scheduler
+
+    *  cause_of_failure
+    *  get_runspec
+    """
+
+    def __init__(self, scheduler):
+        self.scheduler = scheduler
+        self.scheduler.handlers["get_runspec"] = self.get_runspec
+        self.scheduler.handlers["get_error_cause"] = self.get_error_cause
+        self.scheduler.handlers["_process_key"] = self._process_key
+        self.scheduler.extensions["replay-tasks"] = self
+
+    def _process_key(self, key):
+        if isinstance(key, list):
+            key = tuple(key)  # ensure not a list from msgpack
+        key = stringify(key)
+        return key
+
+    def get_error_cause(self, *args, keys=(), **kwargs):
+        for key in keys:
+            key = self._process_key(key)
+            ts = self.scheduler.tasks.get(key)
+            if ts is not None and ts.exception_blame is not None:
+                return ts.exception_blame.key
+
+    def get_runspec(self, *args, key=None, **kwargs):
+        key = self._process_key(key)
+        ts = self.scheduler.tasks.get(key)
+        return {
+            "task": ts.run_spec,
+            "deps": [dts.key for dts in ts.dependencies]
+        }
+
+
+class ReplayTaskClient:
+    """
+    A plugin for the client allowing replay of remote tasks locally
+
+    Adds the following methods (and their async variants)to the given client:
+
+    - ``recreate_error_locally``: main user method for replaying failed tasks
+    - ``recreate_task_locally``: main user method for replaying any task
+    - ``get_futures_error``: gets the task, its details and dependencies,
+        responsible for failure of the given future.
+    - ``get_futures_components``: gets the task, its details and dependencies.
+    """
+
+    def __init__(self, client):
+        self.client = client
+        self.client.extensions["replay-tasks"] = self
+        # monkey patch
+        self.client._get_raw_components_from_future = self._get_raw_components_from_future
+        self.client._prepare_raw_components = self._prepare_raw_components
+        self.client._get_components_from_future = self._get_components_from_future
+        self.client._get_errored_future = self._get_errored_future
+        self.client.recreate_task_locally = self.recreate_task_locally
+        self.client.recreate_error_locally = self.recreate_error_locally
+        self.client._execute_task_components = self._execute_task_components
+
+    @property
+    def scheduler(self):
+        return self.client.scheduler
+
+    async def _get_raw_components_from_future(self, future):
+        await wait(future)
+        # one reason not to pass spec into this function is in case we want to
+        # expose a method to get all the components/deps in the future. This
+        # way it will be easier to do.
```

Comment:
@fjetter - I deviated a bit from your suggestion to reference/fetch the components by spec, for the reason stated above. This comment is referring to a method similar to the current "get_futures_error" which I removed for now for the sake of removing some noise from the PR while restructuring, but can add back. Although I think the recreate_task_locally(run=False) would provide most of that functionality (with the deps resolved instead of raw)

Reply:
I do not mind a different structure. I do care about API surface, though, and would like to only introduce features as needed.

For instance, the `run` argument is something I feel is not necessary. It adds some clutter to the function signature and changes the behaviour of the function significantly. Opposed to returning a tuple with well defined semantics, it instead will execute a function and will return _something_.
For the debugging use case it sounds like the `run=False` is actually what you are looking for. What about just exposing this as part of the API and tell users to call the function themselves?

Reply:
Got it. I'll remove the `run` arg, so that each method only has one way of processing things. I think it would be worth it to make a final decision on the set of publicly exposed methods that we would want before I add any more commits. We have all the pieces now to make it work so deciding on this set is all that should be left.

In my eyes this is the superset of what we could potentially expose (names can be shortened, potentially overly descriptive currently),
```
get_raw_components_from_future - returns func, and raw args/kwargs including raw unresolved future inputs/deps
get_raw_components_from_errored_future - same as above but implicitly calls get_errored_future
                                         equiavelent to current implementation of get_futures_error
get_components_from_future - similar to raw_components, but args/kwargs/deps are resolved/gathered
get_components_from_errored_future - same as above but implicitly calls get_errored_future
get_errored_future - takes a future collection object and returns the future responsible for causing error
recreate_task_locally - runs the code for a future on your local machine
recreate_error_locally - same as above but implicitly calls get_errored_future
```

Personally I think the only things we should expose are: `recreate_task_locally` and `recreate_error_locally`. I feel like the case where you want just the components and not execute the future code is not common for something that's supposed to be a debugging tool, especially since you can inspect the args/kwargs through pdb, when using recreate_*_locally.

Reply:
sounds good to me. Keeping the `recreate_*_locally` methods seems intuitive to me

Reply:
@fjetter - made these changes. also added some analogous recreate_task_locally tests to the existing recreate_error_locally tests (which I also slightly updated to account for private api changes). Also added some preliminary docstrings, which could probably use some work.